### PR TITLE
SAA-1354: Add exclusions to an allocation

### PR DESCRIPTION
**Pull request by smcveigh941**
_**NO BREAKING CHANGES INTRODUCED**_

- Added exclusions to the allocation response model
- Added ability to add exclusions when creating an allocation
- Added ability to add exclusions when updating an allocation
- Added ability to migrate exclusions from nomis

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt [hunk 1]
```diff
@@ -131,6 +133,8 @@ data class ActivitySchedule(
 
   fun slots() = slots.toList()
 
+  fun slots(weekNumber: Int, timeSlot: TimeSlot) = slots().filter { s -> s.weekNumber == weekNumber && s.timeSlot() == timeSlot }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:136]: I am probably missing something here.  Why can this return multiple slots?  Shouldn't there only be one match?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Allocation.kt [hunk 2]
```diff
@@ -87,6 +87,9 @@ data class Allocation(
 
   @Schema(description = "The earliest release date of the prisoner. Included only if includePrisonerSummary = true")
   var earliestReleaseDate: EarliestReleaseDate? = null,
+
+  @Schema(description = "The activity slots which the allocated prisoner is excluded from attending")
+  var exclusions: List<Slot>? = null,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Allocation.kt:92]: Minor but why is this and attributes above `var's`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Exclusion.kt [hunk 3]
```diff
@@ -47,8 +48,41 @@ data class Exclusion(
     DayOfWeek.SATURDAY.takeIf { saturdayFlag },
     DayOfWeek.SUNDAY.takeIf { sundayFlag },
   )
+  fun setDaysOfWeek(days: Set<DayOfWeek>) {
+    if (!activityScheduleSlot.getDaysOfWeek().containsAll(days)) {
+      throw IllegalArgumentException("Cannot set exclusions for slots where the activity does not run")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Exclusion.kt:53]: Perhaps include the schedule id (or slot id) in the message for context/tractability.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt [hunk 4]
```diff
@@ -81,6 +82,14 @@ class AllocationsService(
     }
   }
 
+  private fun applyExclusionsUpdate(request: AllocationUpdateRequest, allocation: Allocation) {
+    request.exclusions?.onEach { exclusion ->
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:86]: Do we need to capture what has changed here so we can raise an allocation amended event?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt [hunk 5]
```diff
@@ -81,6 +82,14 @@ class AllocationsService(
...
+      allocation.activitySchedule.slots(exclusion.weekNumber, exclusion.timeSlot())
+        .apply { if (isEmpty()) throw IllegalArgumentException("No ${exclusion.timeSlot()} slots in week number ${exclusion.weekNumber}") }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:88]: Again I would include an entity id in the exception message for context/traceability.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivityScheduleTest.kt [hunk 6]
```diff
@@ -228,6 +231,59 @@ class ActivityScheduleTest {
...
+            monday = true,
+          ),
+        ),
+        allocatedBy = "FREDDIE",
+      )
+    }
+      .isInstanceOf(IllegalArgumentException::class.java)
+      .hasMessage("No AM slots in week number 3")
+
+    schedule.allocations() hasSize 2
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivityScheduleTest.kt:284]: Perhaps assert that prisoner "654321" is not allocated also? Meh.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt (hunk 1 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:136]: Are you sure? I thought there `could` be multiple slots for an allocation? To cater for same day/different slot AM/PM/ED circumstances.

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt]: ok - so would work for Wk 1, Weds, AM   and Wk1, Weds, PM, and Wk1, Weds, ED - so there could be 3 slots for the activity schedule, all on a Weds, but your function is just finding a unique slot?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt [hunk 7]
```diff
@@ -81,6 +82,14 @@ class AllocationsService(
     }
   }
 
+  private fun applyExclusionsUpdate(request: AllocationUpdateRequest, allocation: Allocation) {
+    request.exclusions?.onEach { exclusion ->
+      allocation.activitySchedule.slot(exclusion.weekNumber, exclusion.timeSlot())
+        .apply { if (this == null) throw IllegalArgumentException("Updating allocation with id ${allocation.allocationId}: No ${exclusion.timeSlot()} slots in week number ${exclusion.weekNumber}") }
+        .let { slot -> allocation.updateExclusion(slot!!, exclusion.getDaysOfWeek()) }
+    }
+  }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:91]: Question - a "Slot" could contain multiple days e.g.  Mon, Tues, Weds - PM. Can we exclude from just the PM slot?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AllocationIntegrationTest.kt [hunk 8]
```diff
@@ -216,6 +219,39 @@ class AllocationIntegrationTest : IntegrationTestBase() {
...
+    }
+
+    verify(eventsPublisher).send(eventCaptor.capture())
+
+    with(eventCaptor.firstValue) {
+      eventType isEqualTo "activities.prisoner.allocation-amended"
+      additionalInformation isEqualTo PrisonerAllocatedInformation(1)
+      occurredAt isCloseTo TimeSource.now()
+    }
+  }
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AllocationIntegrationTest.kt:253]: Might be a few more tests required here to cater for exclusions from multiple slots or multiple days in the same slot? Can  come later though.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateActivityIntegrationTest.kt [hunk 9]
```diff
@@ -349,6 +350,9 @@ class MigrateActivityIntegrationTest : IntegrationTestBase() {
       endDate = null,
       endComment = null,
       suspendedFlag = false,
+      exclusions = listOf(
+        Slot(weekNumber = 1, timeSlot = "AM", monday = true),
+      ),
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/MigrateActivityIntegrationTest.kt:355]: Multiple days set to true? Can come later.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesServiceTest.kt [hunk 10]
```diff
@@ -148,6 +149,21 @@ class ManageAttendancesServiceTest {
...
 
+  @Test
+  fun `attendance is not created if the allocation is excluded`() {
+    instance.activitySchedule.activity.attendanceRequired = true
+
+    allocation.updateExclusion(activitySchedule.slots().first(), setOf(today.dayOfWeek))
+
+    whenever(scheduledInstanceRepository.findAllBySessionDate(today)).thenReturn(listOf(instance))
+    whenever(prisonerSearchApiClient.findByPrisonerNumbers(emptyList())).thenReturn(emptyList())
+
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesServiceTest.kt:160]: Is it the exclusion, or the empty prisoner search results, which determine no attendance is being created?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesServiceTest.kt [hunk 11]
```diff
@@ -467,7 +483,18 @@ class ManageAttendancesServiceTest {
...
+    activity = activityEntity(startDate = activityStartDate, timestamp = activityStartDate.atStartOfDay(), noSchedules = true)
+      .apply {
+        this.addSchedule(
+          activitySchedule(
+            this,
+            activityScheduleId = activityId,
+            activityStartDate.atStartOfDay(),
+            daysOfWeek = setOf(activityStartDate.dayOfWeek),
+            noExclusions = true,
+          ),
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesServiceTest.kt:495]: What does the `noExclusions` param do?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt (hunk 7 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AllocationsService.kt:91]: I worded that question awkwardly - what I meant was can we exclude just the Monday element of the PM slots when multiple days exist, and I think you've answered. Thanks.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AllocationIntegrationTest.kt (hunk 8 above)
Comment (timharrison-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AllocationIntegrationTest.kt:253]: Just making sure the lists/ sets are all applied when multiple exist.

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**